### PR TITLE
Kernel: Allow process with multiple threads to call exec and exit

### PR DESCRIPTION
**Pull request by cbsirb**
This allows a process wich has more than 1 thread to call exec, even
from a thread. This kills all the other threads, but it won't wait for
them to finish, just makes sure that they are not in a running/runable
state.

In the case where a thread does exec, the new program PID will be the
thread TID, to keep the PID == TID in the new process.

This introduces a new function inside the Process class,
kill_threads_except_self which is called on exit() too (exit with
multiple threads wasn't properly working either).

Inside the Lock class, there is the need for a new function,
clear_waiters, which removes all the waiters from the
Process::big_lock. This is needed since after a exit/exec, there should
be no other threads waiting for this lock, the threads should be simply
killed. Only queued threads should wait for this lock at this point,
since blocked threads are handled in set_should_die.

After the do_exec passes the `old_regions = move(m_regions)` part, it's not safe to have any other threads from that process running, they simply crash at any syscall since `MM.validate_user_stack` can't find a region. What I did was to add a new `m_exec_tid`/`exec_tid()` in the Process class, and if that is set, then `Scheduler::pick_next()` will schedule only that thread from the process to run: `if (thread->process().exec_tid() && thread->process().exec_tid() != thread->tid()) continue`

### CODE REVIEW

File: Kernel/Process.cpp [hunk 1]
```diff
@@ -763,19 +763,35 @@ pid_t Process::sys$fork(RegisterState& regs)
...
+
             return IterationDecision::Continue;
         });
-        ASSERT(thread_count() == 1);
-        ASSERT_NOT_REACHED();
     }
 
+    int tries = thread_count() * 4;
+    while (thread_count() > 1 && --tries)
+        Scheduler::yield();
```
Comment (awesomekling) [inline Kernel/Process.cpp]: This will kill all threads in a program that calls `exec()` with bad arguments. In such a case, `exec()` should simply fail and keep the program running like before.

Also, we should do this without having to wait for threads to die. As long as we've marked the threads for death and made them non-runnable, there's no need for us to wait for them to get deallocated.

File: Kernel/Thread.h [hunk 2]
```diff
@@ -264,6 +264,7 @@ class Thread {
     bool is_stopped() const { return m_state == Stopped; }
     bool is_blocked() const { return m_state == Blocked; }
     bool in_kernel() const { return (m_tss.cs & 0x03) == 0; }
+    bool is_dead() const { return m_state == Dead || m_state == Dying; }
```
Comment (awesomekling) [inline Kernel/Thread.h]: `Dead` and `Dying` are two distinct states, so let's not confuse things by having a getter that makes it look like the states are equivalent.

File: Kernel/Process.cpp (hunk 1 above)
Comment (cbsirb) [reply Kernel/Process.cpp]: Yeah, I spent some time to figure things out. I just removed that part and remembered why I put it there: `Queued` threads won't exit (running, blocked, etc. are fine) and they will hit an assertion in `Kernel::Thread::wake_from_queue()`.

Comment (cbsirb) [reply Kernel/Process.cpp]: Figured it out (at least I think I did). I amended the previous commit, now it won't need to actually wait for threads to finish. My only question remains if the `m_joiner` should also be set to null inside the threads. Makes sense to clean it, since the process which does the exec isn't the joiner any more.

Comment (awesomekling) [reply Kernel/Process.cpp]: Sure, nobody is going to join these threads on successful `exec()` so let's clear the `m_joiner`.

Comment (cbsirb) [reply Kernel/Process.cpp]: Set that too. This *should* cover all the cases. In my tests (sleeping threads, running threads, queued threads), it finishes them nicely.

Comment (cbsirb) [reply Kernel/Process.cpp]: Also, I investigated `exit` too, and that doesn't really work if you have queued threads, the program will exit, but it will just hang the terminal.

I pushed a working solution for both exec and exit & updated the PR.

File: Kernel/Process.cpp [hunk 3]
```diff
@@ -760,21 +760,44 @@ pid_t Process::sys$fork(RegisterState& regs)
...
-        });
-        ASSERT(thread_count() == 1);
-        ASSERT_NOT_REACHED();
-    }
+
+    kill_threads_except_self();
+
+    // Give the other threads a chance to exit
+    if (m_thread_count > 1)
+        Scheduler::yield();
```
Comment (awesomekling) [inline Kernel/Process.cpp]: Why is this necessary?

File: Kernel/Process.cpp [hunk 4]
```diff
@@ -1183,6 +1209,17 @@ int Process::exec(String path, Vector<String> arguments, Vector<String> environm
     else if (elf_result.error().is_error())
         return elf_result.error();
 
+    if (set_in_exec()) {
```
Comment (awesomekling) [inline Kernel/Process.cpp]: Why is this necessary? Why would we re-enter `do_exec`?

File: Kernel/Process.cpp [hunk 5]
```diff
@@ -760,21 +760,44 @@ pid_t Process::sys$fork(RegisterState& regs)
...
-        dbgprintf("Gonna die because I have many threads! These are the threads:\n");
-        for_each_thread([](Thread& thread) {
-            dbgprintf("Thread{%p}: TID=%d, PID=%d\n", &thread, thread.tid(), thread.pid());
-            return IterationDecision::Continue;
-        });
-        ASSERT(thread_count() == 1);
-        ASSERT_NOT_REACHED();
-    }
+
+    kill_threads_except_self();
```
Comment (awesomekling) [inline Kernel/Process.cpp]: The FIXME that you are removing says:
>Kill any threads the moment we commit to the exec().

At this point, we have not yet committed to the exec! It can still fail and return with an error.
In those cases, we should not kill any threads.

There's a comment further down where we actually commit:
```
        // NOTE: At this point, we've committed to the new executable.
```

File: Kernel/Process.cpp (hunk 3 above)
Comment (cbsirb) [reply Kernel/Process.cpp]: It isn't any more, it was there to fix a crash I had, but in the new version where we kill the threads only after committing to exec it's removed.

File: Kernel/Process.cpp (hunk 4 above)
Comment (cbsirb) [reply Kernel/Process.cpp]: We won't re-enter exec, but maybe a thread calls exec() at the same time as another (and I used to get a crash when that happened). But I removed it anyway, in the last version, fixed it another way (I'll comment in the PR directly)

File: Kernel/Process.cpp (hunk 5 above)
Comment (cbsirb) [reply Kernel/Process.cpp]: Did it there because I used to have a crash, but I commented at the end of the PR more details.

We can be interrupted (and it happens every time with >100 threads) in `do_exec` after the `move(m_regions)` part, and any other thread running from that point will crash if a syscall happens. Fixed it another way (see the new push).

File: Kernel/Process.cpp [hunk 6]
```diff
@@ -3041,14 +3065,10 @@ void Process::die()
...
-        InterruptDisabler disabler;
-        for_each_thread([](Thread& thread) {
-            thread.set_should_die();
-            return IterationDecision::Continue;
-        });
-    }
+    Thread::current->set_should_die();
+
+    // Tell the other threads to unwind and die.
+    kill_threads_except_self();
```
Comment (awesomekling) [inline Kernel/Process.cpp]: `Process::die()` may be called on a not-currently-running process during signal dispatch.
This logic would not work in that case, since this assumes that `Thread::current` is a thread owned by `this`.

File: Kernel/Process.cpp [hunk 7]
```diff
@@ -991,6 +1010,7 @@ int Process::do_exec(NonnullRefPtr<FileDescription> main_program_description, Ve
     m_master_tls_size = master_tls_size;
     m_master_tls_alignment = master_tls_alignment;
 
+    m_pid = new_main_thread->tid();
```
Comment (awesomekling) [inline Kernel/Process.cpp]: How do other systems behave? It seems weird that the result of `getpid()` would change after `exec()`.

Comment (cbsirb) [reply Kernel/Process.cpp]: Linux has the concept of `group_leader`, which in serenity would be the `new_main_thread`. The `group_leader` always has the pid == tid, that's why I did it this way. 

But just noticed that in Linux it's the other way around, it should have been `Thread::current->m_tid = m_pid`, but this isn't possible, since the old `main_thread` may have not exited yet.

Comment (cbsirb) [reply Kernel/Process.cpp]: > Before exec: pid: 8428, tid: 8429
> After  exec: pid: 8428, tid: 8428

This is how Linux behaves with a thread that does exec, it just borrows the group_leader tid.

File: Kernel/Process.cpp (hunk 6 above)
Comment (cbsirb) [reply Kernel/Process.cpp]: Indeed, this was a mistake. I solved in, now we just kill all the process threads, we don't use `Thread::current` any more.

**Comment by awesomekling** [comment]
Oh, this is looking really nice :)

**Comment by awesomekling** [comment]
Oops, I forgot to press the merge button yesterday! Thanks for reminding me, and thank you for fixing this :D
